File: src/test/baekjoon/java/dfsbfs/VillageCount.java
package test.baekjoon.java.dfsbfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class VillageCount {
    static int answer = -1;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
        boolean[] visited = new boolean[101];
        int N = Integer.parseInt(br.readLine());

        for(int i = 0; i < N + 1; i++) {
            graph.add(new ArrayList<>());
        }

        StringTokenizer st;
        st = new StringTokenizer(br.readLine());
        int findX = Integer.parseInt(st.nextToken());
        int findY = Integer.parseInt(st.nextToken());

        int M = Integer.parseInt(br.readLine());
        for(int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            graph.get(x).add(y);
            graph.get(y).add(x);
        }

        dfs(graph, visited, findX, findY, 0);

        System.out.println(answer);
    }

    static void dfs(ArrayList<ArrayList<Integer>> graph, boolean[] visited, int pos, int find, int count) {
        visited[pos] = true;

        for(int i : graph.get(pos)) {
            if(!visited[i]) {
                if(i == find) {
                    answer = count + 1;
                    return;
                }

                dfs(graph, visited, i, find, count + 1);
            }
        }
    }
}

File: src/test/baekjoon/python/dfsbfs/VillageCount.py
N = int(input())
find_x, find_y = map(int, input().split())
graph = [list([]) for _ in range(101)]
visited = [False] * 101
M = int(input())
for _ in range(M):
    x, y = map(int, input().split())
    graph[x].append(y)
    graph[y].append(x)

answer = -1
def dfs(graph, visited, pos, find, count):
    visited[pos] = True

    for i in graph[pos]:
        if visited[i] is False:
            if i == find:
                global answer
                answer = count + 1
                return
            dfs(graph, visited, i, find, count + 1)

dfs(graph, visited, find_x, find_y, 0)
print(answer)
